### ISSUE TITLE
DKIM signing from trusted networks

### ISSUE THREAD

**Issue by foogod** [vstakhov/rmilter#139]
Hiho..  I have rmilter set up and working well with postfix except for one small issue:

It will correctly DKIM-sign outbound messages when the sender has authenticated (via SASL) with the postfix server, but it does not sign any messages sent by local processes on the host itself (which just invoke postfix directly).  I have also noticed it does not sign messages sent (without user authentication) from other hosts on networks in the 'mynetworks' list, despite the fact that these are supposed to be trusted networks from which outbound relaying is otherwise allowed.

Is there a way to configure rmilter to perform DKIM signing for unauthenticated connections from local processes and trusted networks (so that all hosts which have relaying allowed also get all their outbound messages signed)?

**Comment by xpunkt** [comment]
# /etc/postfix/main.cf
milter_rmilter = inet:[127.0.0.1]:8894

non_smtpd_milters = $milter_rmilter
smtpd_milters = $milter_rmilter

milter_protocol = 6
milter_mail_macros = i {mail_addr} {client_addr} {client_name} {auth_auten}
milter_default_action = accept

just to be sure you have it like i have, but there is 2 questions you ask about:

one it you like to sign all in network
another when sasl authed

reading how you describe it, it works as designed, but the on to sign all in trusted networks is not imho supported with rmilter

opendkim can do this, i think just it would make more sense to make rmilter support ATPS signing, opendkim can sign all, buts its not ATPS

**Comment by fatalbanana** [comment]
You can set `sign_networks` in `dkim` config section to list of addresses/networks you want to do signing for irrespective of auth, or set `auth_only` to `no`.

**Comment by foogod** [comment]
Thanks, I had completely missed the `sign_networks` option.. that does fix the issue of mail relayed from other hosts on trusted subnets..

Unfortunately, it does not appear to help for mail sent by processes on the mail server host itself.  Is there any way to get rmilter to consider "local" process submission to be trusted/authenticated for DKIM signing?

(I don't want to set `auth_only` to `no`, because this server is also used for inbound mail and that would potentially allow spoofers sending mail inbound from untrusted sources to get my server to sign their emails too..)

**Comment by vstakhov** [comment]
Why not add local addresses to the `sign_network` list? If you need more flexible DKIM signing I would recommend DKIM signatures from Rspamd: https://rspamd.com/doc/modules/dkim.html#dkim-signatures

They are natively supported by Rmilter, so you can use Lua script and Rspamd Lua API to write whatever logic you like.

**Comment by foogod** [comment]
I had already added 127.0.0.0/8 and [::1]/128 to `sign_networks`.  This works fine for clients which actually connect via SMTP to localhost, but does not work for processes which submit mail by invoking the postfix/sendmail command directly, as those do not use a TCP connection.

I was not aware that rspamd would do DKIM-signing as well.. It does look a lot more customizable, so I'll definitely look into that in more detail when I have a bit of time.  The same question remains, though:  How to have the Lua script be able to recognize mail that was submitted via a local command/unix-socket?

**Comment by xpunkt** [comment]
postconf myorigin domain miss in rmilter ?

**Comment by foogod** [comment]
So I feel kinda dumb here..

I tried building and installing the 2b86b22 version, and unfortunately it didn't seem to help, then I eventually noticed in my mail logs that, for locally-submitted mail, rmilter didn't even appear to be being invoked at all.  From this I realized that my Postfix config had a `smtp_milters` setup for rmilter, but no corresponding `non_smtpd_milters` line.  Adding the following to my Postfix `main.cf` immediately fixed the problem:

    non_smtpd_milters = inet:127.0.0.1:11332

I checked and this actually works for the stock 1.9.1 version of rmilter as well (when invoking rmilter from `non_smtpd_milters` postfix apparently pretends it's from 127.0.0.1 already), so it looks like the change in 2b86b22 isn't actually needed at all (and doesn't change anything one way or the other).

Sorry about wasting everyone's time.. looks like this was all just a postfix misconfiguration issue on my part..
